### ISSUE TITLE
support providing `project-name` via hooks, env vars and template value file

### ISSUE THREAD

**Issue by xoac** [cargo-generate/cargo-generate#522]
Hi, 
I wanted to create a template that allows me to create a project within URL. The question about `project-name` in most cases is in GitHub link and this is in my opinion, the best approach in creating a new project from template. The `project-name` variable must be required, but checking about its absence should be postponed until all pre-hooks are called.

This is a pseudo dialogue I want to achieve:

**Past link to repository of your project:** <github.com/organizaton/my-awesome-rust-project>
**Is the name of your project correct `my-awesome-rust-project`?** [Y (Yes)/e (Edit)] 
**Your project is created!!**

To make this backward compatible the template should set options that allow it to set `project-name` on it's own. For example in `cargo-generate.toml` `template.project-name.ask = false` (default `true`)

What do you think about this idea?

**Comment by sassman** [comment]
Interesting feature. 

Let me explain 2 arguments why I think that would be a bit counter productive.

1. all my templates have the name "template" in the url, and I don't want to be have the project to be called template.
2. `cargo-generate` supports to have multiple sorts of templates in one single git repository. For example you can have a folder `cli` and a folder `lib` in the repository. Then when using `cargo-generate` with this one repository url it depends on the a different command line option what folder will be taken. So therefore the repository URL would also not reflect the right project name.

**Comment by xoac** [comment]
I think you misunderstood me, I suggest:
1. The cargo-generate require `{project-name}` to be defined (but not exactly as user input).
2. template is allowed to disable prompt for `{project-name}`
3. cargo-generate will exit with error if template will not provide `{project-name}` variable with pre-hook scripts.

Maybe examples would be more clear:
1. You want to create `randomly-named-rust-project-template`
2. cargo-generate ask about `{project-name}` (a bit ironic...)
3. You ignore user input and overwrite it with pre-hook script with random name for project

In my case I want to create template like this:
1. You run `cargo generate --git me/my-template`
2. You are asked about git origin of project  (so `github.com/me/my-awesome-project`)
3. `my-template` sets `{project-name}` to `my-awesome-project`

**Comment by sassman** [comment]
Yeah, I have misunderstood you in the first place.

Could you please show me such a template that would provide the `project-name` via hook?

In general it make sense that the `project-name` is also possible to provide via
- environment variable: `CARGO_GENERATE_VALUE_PROJECT_NAME`
- a [template value file](https://github.com/cargo-generate/cargo-generate#---template_values_file-flag)
- or finally via hook

However, there are a few things that needs to be kept in mind. Since the project-name is used for a folder name and a crate name, it must be strictly validated. It bares a security risk to blindly allow any folder name that a template provides. 

Having that said, we have basically 2 cases:
1. the user of the template is in charge of the name (env var or template value file)
2. the template author is in charge of the name (hooks)

So I tend to be sceptical that supporting case 2 is a wise thing to do (from a security point of view).

**Comment by xoac** [comment]
The level of security can be unchanged. The only thing that would change will be the point when `{project-name}` and variables that depends on `{project-name}` will be validated by cargo-generate. 

1. Ask about all variables defined by `cargo-geneate.toml` 
2. Run pre-hooks that allow modify variables or add new
3. Validate variables required by cargo-genare (like `project_dir`, `project-name`, `crate_name`,  `crate_type`)

So the validation should be done before this line: https://github.com/cargo-generate/cargo-generate/blob/69914f177317b6cc3f02a8d08bdd0e958f6f677c/src/lib.rs#L116

----

At this moment I am able to provide invalid value to `project-name`

CLI:
```
cargo generate -v --path ~/projects/github.com/xaoc/opinionated-rust-template/ template --name invalid2
```
and code in pre-hook that set invalid `project-name`
```
variable::set("project-name", "🦞");
```
and head of Cargo.toml:
```
[package]
name = "🦞"
version = "0.1.0"
```
(this will not affect `project_dir` since this variable is evaluate in rust code before calling pre-hooks. )

-----

> Could you please show me such a template that would provide the project-name via hook?

I want to write one so any working example yet.

**Comment by sassman** [comment]
Thanks for elaboration, sounds consistent to me.

> I want to write one so any working example yet.

Looking forward to have such a template to simplify development efforts.

**Comment by xoac** [comment]
I have published example template that should provide `project-name` by hooks. Here is the link: https://github.com/xoac/opinionated-rust-template

**Comment by taurr** [comment]
I'll give this a go after #669 has been merged - otherwise I'll invite the merge conflict gods!

**Comment by taurr** [comment]
@xoac, @sassman  I have a few issues, actually its kinda breaking...

1. Currently, the pre-hook might actually depend on the project-name already being set.
If `cargo-generate` waits with asking until after the pre-hook, any templates that uses the variables in question from the pre-hook will break! (granted, it's not many!)

2. If we delay the check and ask a little later, it should really include _all_ the builtin variables.
